avalanche/training/plugins/generative_replay.py

```diff
@@ -0,0 +1,157 @@
+                            **kwargs):
+        """
+        Make deep copies of generator and solver before training new experience.
+        """
+        if self.untrained_solver:
+            # The solver needs to be trained before labelling generated data and
+            # the generator needs to be trained before we can sample.
+            return
+        self.old_generator = deepcopy(self.generator)
+        self.old_generator.eval()
+        if not self.model_is_generator:
+            self.old_model = deepcopy(strategy.model)
+            self.old_model.eval()
+
+    def after_training_exp(self, strategy: "SupervisedTemplate",
+                           num_workers: int = 0, shuffle: bool = True,
+                           **kwargs):
+        """
+        Set untrained_solver boolean to False after (the first) experience,
+        in order to start training with replay data from the second experience.
+        """
+        self.untrained_solver = False
+
+    def before_training_iteration(self, strategy: "SupervisedTemplate",
+                                  **kwargs):
+        """
+        Generating and appending replay data to current minibatch before 
+        each training iteration.
+        """
+        if self.untrained_solver:
+            # The solver needs to be trained before labelling generated data and
+            # the generator needs to be trained before we can sample.
+            return
+        # extend X with replay data
+        replay = self.old_generator.generate(
+            len(strategy.mbatch[0]) * (strategy.experience.current_experience)
+            ).to(strategy.device)  
+        strategy.mbatch[0] = torch.cat([strategy.mbatch[0], replay], dim=0)
+        # extend y with predicted labels (or mock labels if model==generator)
+        if not self.model_is_generator:
+            with torch.no_grad():
+                replay_output = self.old_model(replay).argmax(dim=-1)
+        else:
+            # Mock labels:
+            replay_output = torch.zeros(replay.shape[0])
+        strategy.mbatch[1] = torch.cat(
+            [strategy.mbatch[1], replay_output.to(strategy.device)], dim=0)
+        # extend task id batch (we implicitley assume a task-free case)
+        strategy.mbatch[-1] = torch.cat([strategy.mbatch[-1], torch.ones(
+            replay.shape[0]).to(strategy.device) * strategy.mbatch[-1][0]],
+             dim=0)
+
+
+class TrainGeneratorAfterExpPlugin(SupervisedPlugin):
+    """
+    TrainGeneratorAfterExpPlugin makes sure that after each experience of 
+    training the solver of a scholar model, we also train the generator on the 
+    data of the current experience.
+    """
+
+    def after_training_exp(self, strategy: "SupervisedTemplate", **kwargs):
+        """
+        The training method expects an Experience object 
+        with a 'dataset' parameter.
+        """
+        strategy.plugins[1].generator_strategy.train(strategy.experience) 
```

Comment:
How do you know that the `GenerativeReplayPlugin` has index 1 in the plugins list? 
Generally you cannot make this assumption!

You can cycle throughout all the plugins and select the one you interested with using something like 
```python
for plugin in strategy.plugins:
    if type(plugin) is GenerativeReplayPlugin:
        plugin.generator_strategy.train(strategy.experience) 
```

Reply:
Perfect, thank you! Looks a lot more robust now :+1: